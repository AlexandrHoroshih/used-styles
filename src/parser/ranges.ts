export interface LocalCodeLocation {
  line: number;
  column: number;
}

// tslint:disable-next-line:no-empty-interface
export type CodeLocation = LocalCodeLocation;

export interface CodeLocationRange {
  start: CodeLocation;
  end: CodeLocation;
}

export const rangesEqual = (a: CodeLocation, b: CodeLocation) =>
  // a.file === b.file &&
  a.line === b.line && a.column === b.column;

export const rangesIntervalEqual = (a: CodeLocationRange, b: CodeLocationRange) =>
  rangesEqual(a.start, b.start) && rangesEqual(a.end, b.end);

export const localRangeMin = (v: CodeLocation, max: LocalCodeLocation): CodeLocation => {
  if (v.line < max.line) {
    return v;
  }

  if (v.line === max.line) {
    return {
      // file: v.file,
      line: v.line,
      column: Math.min(v.column, max.column),
    };
  }

  return {
    // file: v.file,
    ...max,
  };
};

export const localRangeMax = (v: CodeLocation, min: LocalCodeLocation): CodeLocation => {
  if (v.line > min.line) {
    return v;
  }

  if (v.line === min.line) {
    return {
      // file: v.file,
      line: v.line,
      column: Math.max(v.column, min.column),
    };
  }

  return {
    // file: v.file,
    ...min,
  };
};

export const createRange = (line: number, column: number): CodeLocation => ({
  line,
  column,
});
