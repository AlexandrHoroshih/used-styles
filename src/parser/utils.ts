const classish = (str: string): boolean => !!str && str.indexOf('.') >= 0;

export const mapStyles = (styles: string) =>
  (
    styles
      // remove style body
      .replace(/({[^{}]+})/g, '$')
      .replace(/({[^{}]+})/g, '$')
      // match style name
      .match(/\.([^>~,+$:{\[\s]+)?/g) || []
  )
    // clean style name
    .map(x => x.replace(/[\s,.>~+$]+/, ''))
    .map(x => x.replace(/[.\s.:]+/, ''));

export const mapSelector = (selector: string) => {
  const ruleSelection =
    // anything like "style"
    selector.match(/\.([^>~+$:{\[\s]+)?/g) || [];

  ruleSelection.reverse();

  const effectiveMatcher: string = ruleSelection.find(classish) || '';
  const selectors = effectiveMatcher.match(/(\.[^.>~+,$:{\[\s]+)?/g);

  return (selectors || []).map(x => x.replace(/[.\s.:]+/, '')).filter(Boolean);
};
