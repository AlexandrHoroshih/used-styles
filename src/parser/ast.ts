import { CodeLocation } from './ranges';

export interface StyleRule {
  prop: string;
  value: string;
  important: boolean;
}

export interface ProcessedAtRule {
  kind: 'media' | 'layer';
  value: string;
}

export interface StyleSelector {
  selector: string;
  postfix: string;

  pieces: string[];
  atrules: ProcessedAtRule[];
  parents?: string[];

  declaration: number;
  hash: string;
}

export interface StyleBody {
  id: number;
  rules: StyleRule[];
  start: CodeLocation;
  end: CodeLocation;
}

export type StyleBodies = Record<number, StyleBody>;

export interface AtRule {
  kind: string;
  id: string;
  css: string;
}

export type AtRules = AtRule[];

export interface SingleStyleAst {
  file: string;
  selectors: StyleSelector[];
  bodies: StyleBodies;
  atRules: AtRules;
}

export type StyleAst = Record<string, SingleStyleAst>;
