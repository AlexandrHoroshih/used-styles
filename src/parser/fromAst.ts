import { kashe } from 'kashe';
import * as postcss from 'postcss';

import { SelectionFilter } from '../types';
import { ProcessedAtRule, SingleStyleAst, StyleBody, StyleRule, StyleSelector } from './ast';

const separator = process.env.NODE_ENV === 'production' ? '' : '\n';

export const escapeValue = (value: string, name: string) => {
  if (name === 'content') {
    return value.split('\\').join('\\\\');
  }

  return value;
};

const createDecl = (decl: StyleRule) => postcss.decl(decl) + ';';

const declsToString = (rules: StyleRule[]) => rules.map((decl) => createDecl(decl)).join(separator);

const getProcessedAtRules = ({ atrules }: { atrules: ProcessedAtRule[] }) => {
  const prefix: string[] = [];
  const postfix: string[] = [];

  atrules.forEach((currentRule) => {
    prefix.push(`@${currentRule.kind} ${currentRule.value} {`);
    postfix.push('}');
  });

  return [prefix.join(separator), postfix.join(separator)];
};

type SelectorLookUp = Set<string>;

const renderRule = kashe(
  (rule: StyleSelector, style: StyleBody) => `${rule.selector} { ${declsToString(style.rules)} }`
);

const isMatching = (rule: StyleSelector, rules: SelectorLookUp) =>
  rule.pieces.length > 0 && rule.pieces.every((piece) => rules.has(piece));

const findMatchingSelectors = (rules: SelectorLookUp, selectors: StyleSelector[]): StyleSelector[] =>
  selectors.filter((rule) => isMatching(rule, rules));

const findUnmatchableSelectors = (selectors: StyleSelector[]): StyleSelector[] =>
  selectors.filter((rule) => rule.pieces.length === 0);

export const fromAst = (rules: string[], def: SingleStyleAst, filter?: SelectionFilter) => {
  const blocks: StyleSelector[] = [];
  const lookup: SelectorLookUp = new Set(rules);

  blocks.push(
    ...findMatchingSelectors(lookup, def.selectors).filter((block) => !filter || filter(block.selector, block))
  );

  return convertToString(blocks, def);
};

export const getUnmatchableRules = (def: SingleStyleAst, filter?: SelectionFilter): StyleSelector[] =>
  findUnmatchableSelectors(def.selectors).filter((block) => !filter || filter(block.selector, block));

export const extractUnmatchable = (def: SingleStyleAst, filter?: SelectionFilter) =>
  convertToString(getUnmatchableRules(def, filter), def) + getAtRules(def);

const getAtRules = (def: SingleStyleAst) => def.atRules.reduce((acc, rule) => acc + rule.css, '');

export const convertToString = (blocks: StyleSelector[], { bodies }: SingleStyleAst) => {
  blocks.sort((ruleA, ruleB) => bodies[ruleA.declaration].id - bodies[ruleB.declaration].id);

  const result: string[] = [];

  let lastMedia = ['', ''];

  blocks.forEach((block, index) => {
    const media = getProcessedAtRules(block);

    if (media[0] !== lastMedia[0]) {
      result.push(lastMedia[1]);
      lastMedia = media;
      result.push(lastMedia[0]);
    }

    if (index < blocks.length - 1 && block.declaration === blocks[index + 1].declaration) {
      result.push(`${block.selector},`);
    } else {
      result.push(renderRule(block, bodies[block.declaration]));
    }
  });

  result.push(lastMedia[1]);

  return result.join(separator);
};
