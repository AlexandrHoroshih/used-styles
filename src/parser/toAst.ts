// @ts-ignore
import * as crc32 from 'crc-32';
import * as postcss from 'postcss';
import { AtRule, Rule } from 'postcss';

import { AtRules, SingleStyleAst, StyleBodies, StyleBody, StyleSelector } from './ast';
import { createRange, localRangeMax, localRangeMin, rangesIntervalEqual } from './ranges';
import { mapSelector } from './utils';

const getAtRule = (rule: AtRule | Rule): string[] => {
  if (rule && rule.parent && 'name' in rule.parent && rule.parent.name === 'media') {
    return getAtRule(rule.parent as any).concat(rule.parent.params);
  }

  return [];
};

const getBreak = (rule: string) => {
  const breakPoints = [
    rule.indexOf(' '),
    rule.indexOf('>'),
    rule.indexOf('~'),
    rule.indexOf('+'),
    rule.indexOf(':'),
  ].filter((index) => index > 0);

  if (breakPoints.length === 0) {
    return rule.length;
  }

  const min = Math.min(...breakPoints);

  return min ? min : rule.length;
};

const getBeforePostfix = (rule: string) => {
  return rule.substr(0, getBreak(rule)).trim();
};

const getPostfix = (rule: string) => {
  return rule.substr(getBreak(rule)).trim();
};

let bodyCounter = 1;

const assignBody = (decl: StyleBody, bodies: StyleBodies): StyleBody => {
  const d = Object.values(bodies).find((bodyDecl) => rangesIntervalEqual(bodyDecl, decl));

  if (d) {
    return d;
  }

  decl.id = bodyCounter++;
  bodies[decl.id] = decl;

  return decl;
};

const hashString = (str: string) => {
  return crc32.str(str).toString(32);
};

const hashBody = (body: StyleBody) => {
  return hashString(JSON.stringify(body.rules));
};

export const buildAst = (CSS: string, file = ''): SingleStyleAst => {
  const root = postcss.parse(CSS);
  const selectors: StyleSelector[] = [];
  const atRules: AtRules = [];

  const bodies: StyleBodies = {};

  const atParents = new Set<any>();

  root.walkAtRules((rule) => {
    if (rule.name === 'charset') {
      return;
    }

    if (rule.name !== 'media') {
      atParents.add(rule);

      atRules /*[rule.params]*/
        .push({
          kind: rule.name,
          id: rule.params,
          css: rule.toString(),
        });
    }
  });

  root.walkRules((rule) => {
    if (atParents.has(rule.parent)) {
      return;
    }

    const ruleSelectors = rule.selector.split(',');

    ruleSelectors
      .map((sel) => sel.trim())
      .forEach((selector) => {
        const stand: StyleSelector = {
          media: getAtRule(rule),
          selector,
          pieces: mapSelector(getBeforePostfix(selector)),
          postfix: getPostfix(selector),
          declaration: 0,
          hash: selector,
        };

        const delc: StyleBody = {
          id: NaN,
          rules: [],
          start: createRange(Infinity, Infinity),
          end: createRange(0, 0),
        };

        rule.walkDecls(({ prop, value, source, important }) => {
          if (source) {
            delc.start = localRangeMin(delc.start, source.start!);
            delc.end = localRangeMax(delc.end, source.end!);

            delc.rules.push({
              prop,
              value,
              important,
            });
          }
        });

        stand.declaration = assignBody(delc, bodies).id;
        stand.hash = `${selector}${hashBody(delc)}${hashString(stand.postfix)}${hashString(stand.media.join())}`;

        selectors.push(stand);
      });
  });

  return {
    file,
    selectors,
    bodies,
    atRules,
  };
};
