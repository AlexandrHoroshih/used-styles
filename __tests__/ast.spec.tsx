import { fromAst } from '../src/parser/fromAst';
import { buildAst } from '../src/parser/toAst';

describe('test ast', () => {
  describe('smoke', () => {
    const CSS = `
      .a,
      .b .c {
        border:1px solid;
        margin: 6px 10px;
      }
      
      .d ~ .e:not(focused){ 
        display: block !important;
        position: relative;
        width: calc(100% - 10px);
      }
      
      @media only screen and (max-width: 600px) {
        .c { position: absolute; }
        .a { position: relative; }  
      }
      
      .a, .b, input { color: rightColor }
      
      .dou.ble {
        color: red
      }
    `;

    it('should map simple style', () => {
      const ast = buildAst(CSS);
      expect(ast).toMatchSnapshot();
    });

    it('should remap simple style', () => {
      const ast = buildAst(CSS);
      const css = fromAst(['e'], ast);
      expect(css).toMatchSnapshot();
    });

    it('dont map unused styles', () => {
      const ast = buildAst(CSS);
      const cssFalsePositive = fromAst(['d', 'b'], ast);
      expect(cssFalsePositive).toBe('.b { color: rightColor; }\n');
    });

    it('handles double classes', () => {
      const ast = buildAst(CSS);
      const single = fromAst(['ble'], ast);
      expect(single).toBe('');

      const double = fromAst(['ble', 'dou'], ast);
      expect(double).toBe('.dou.ble { color: red; }\n');
    });

    it('should remap complex style', () => {
      const ast = buildAst(CSS);
      const css = fromAst(['a', 'c'], ast);
      expect(css).toMatchSnapshot();
    });
  });

  it('should handle keyframes and media', () => {
    const CSS = `
    @-webkit-keyframes ANIMATION_NAME{
      0%{-webkit-transform:rotate(0deg);}
      to{-webkit-transform:rotate(359deg);}}
    @keyframes ANIMATION_NAME{
      0%{-webkit-transform:rotate(0deg);}
      to{-webkit-transform:rotate(359deg);}
    }`;

    const ast = buildAst(CSS);
    expect(ast).toMatchSnapshot();
  });
});
