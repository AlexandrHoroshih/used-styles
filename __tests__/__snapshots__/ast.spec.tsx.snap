// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ast should handle keyframes and media 1`] = `
Object {
  "atRules": Array [
    Object {
      "css": "@-webkit-keyframes ANIMATION_NAME{
      0%{-webkit-transform:rotate(0deg);}
      to{-webkit-transform:rotate(359deg);}}",
      "id": "ANIMATION_NAME",
      "kind": "-webkit-keyframes",
    },
    Object {
      "css": "@keyframes ANIMATION_NAME{
      0%{-webkit-transform:rotate(0deg);}
      to{-webkit-transform:rotate(359deg);}
    }",
      "id": "ANIMATION_NAME",
      "kind": "keyframes",
    },
  ],
  "bodies": Object {},
  "file": "",
  "selectors": Array [],
}
`;

exports[`test ast smoke should map simple style 1`] = `
Object {
  "atRules": Array [],
  "bodies": Object {
    "1": Object {
      "end": Object {
        "column": 25,
        "line": 5,
      },
      "id": 1,
      "rules": Array [
        Object {
          "important": undefined,
          "prop": "border",
          "value": "1px solid",
        },
        Object {
          "important": undefined,
          "prop": "margin",
          "value": "6px 10px",
        },
      ],
      "start": Object {
        "column": 9,
        "line": 4,
      },
    },
    "2": Object {
      "end": Object {
        "column": 33,
        "line": 11,
      },
      "id": 2,
      "rules": Array [
        Object {
          "important": true,
          "prop": "display",
          "value": "block",
        },
        Object {
          "important": undefined,
          "prop": "position",
          "value": "relative",
        },
        Object {
          "important": undefined,
          "prop": "width",
          "value": "calc(100% - 10px)",
        },
      ],
      "start": Object {
        "column": 9,
        "line": 9,
      },
    },
    "3": Object {
      "end": Object {
        "column": 32,
        "line": 15,
      },
      "id": 3,
      "rules": Array [
        Object {
          "important": undefined,
          "prop": "position",
          "value": "absolute",
        },
      ],
      "start": Object {
        "column": 14,
        "line": 15,
      },
    },
    "4": Object {
      "end": Object {
        "column": 32,
        "line": 16,
      },
      "id": 4,
      "rules": Array [
        Object {
          "important": undefined,
          "prop": "position",
          "value": "relative",
        },
      ],
      "start": Object {
        "column": 14,
        "line": 16,
      },
    },
    "5": Object {
      "end": Object {
        "column": 39,
        "line": 19,
      },
      "id": 5,
      "rules": Array [
        Object {
          "important": undefined,
          "prop": "color",
          "value": "rightColor",
        },
      ],
      "start": Object {
        "column": 23,
        "line": 19,
      },
    },
    "6": Object {
      "end": Object {
        "column": 18,
        "line": 22,
      },
      "id": 6,
      "rules": Array [
        Object {
          "important": undefined,
          "prop": "color",
          "value": "red",
        },
      ],
      "start": Object {
        "column": 9,
        "line": 22,
      },
    },
  },
  "file": "",
  "selectors": Array [
    Object {
      "declaration": 1,
      "hash": ".a-1sc6hdk00",
      "media": Array [],
      "pieces": Array [
        "a",
      ],
      "postfix": "",
      "selector": ".a",
    },
    Object {
      "declaration": 1,
      "hash": ".b .c-1sc6hdk-1gpll6f0",
      "media": Array [],
      "parents": Array [
        "b",
      ],
      "pieces": Array [
        "c",
      ],
      "postfix": ".c",
      "selector": ".b .c",
    },
    Object {
      "declaration": 2,
      "hash": ".d ~ .e:not(focused)-1u2ggaf-o0go610",
      "media": Array [],
      "parents": Array [
        "d",
      ],
      "pieces": Array [
        "e",
      ],
      "postfix": "~ .e:not(focused)",
      "selector": ".d ~ .e:not(focused)",
    },
    Object {
      "declaration": 3,
      "hash": ".c-13utu1a0-lnglh",
      "media": Array [
        "only screen and (max-width: 600px)",
      ],
      "pieces": Array [
        "c",
      ],
      "postfix": "",
      "selector": ".c",
    },
    Object {
      "declaration": 4,
      "hash": ".a18suomu0-lnglh",
      "media": Array [
        "only screen and (max-width: 600px)",
      ],
      "pieces": Array [
        "a",
      ],
      "postfix": "",
      "selector": ".a",
    },
    Object {
      "declaration": 5,
      "hash": ".a-1o1j63n00",
      "media": Array [],
      "pieces": Array [
        "a",
      ],
      "postfix": "",
      "selector": ".a",
    },
    Object {
      "declaration": 5,
      "hash": ".b-1o1j63n00",
      "media": Array [],
      "pieces": Array [
        "b",
      ],
      "postfix": "",
      "selector": ".b",
    },
    Object {
      "declaration": 5,
      "hash": "input-1o1j63n00",
      "media": Array [],
      "pieces": Array [],
      "postfix": "",
      "selector": "input",
    },
    Object {
      "declaration": 6,
      "hash": ".dou.ble570uae00",
      "media": Array [],
      "pieces": Array [
        "dou",
        "ble",
      ],
      "postfix": "",
      "selector": ".dou.ble",
    },
  ],
}
`;

exports[`test ast smoke should remap complex style 1`] = `
".a,
.b .c { border: 1px solid;
margin: 6px 10px; }

@media only screen and (max-width: 600px) {
.c { position: absolute; }
.a { position: relative; }
}

.a { color: rightColor; }
"
`;

exports[`test ast smoke should remap simple style 1`] = `
".d ~ .e:not(focused) { display: block !important;
position: relative;
width: calc(100% - 10px); }
"
`;
